source/mqtt_client_connection.c

```diff
@@ -555,85 +1003,60 @@ napi_value mqtt_client_connection_publish(napi_env env, napi_callback_info info)
-        napi_delete_reference(env, metadata->on_publish);
-        napi_async_destroy(env, metadata->on_publish_ctx);
+    if (context->callback) {
+        napi_delete_reference(env, context->callback);
+        napi_async_destroy(env, context->callback_ctx);
     }
 
-    aws_mem_release(allocator, metadata);
+    aws_mem_release(allocator, context);
 
     return NULL;
 }
 
 /*******************************************************************************
  * Subscribe
  ******************************************************************************/
-
-struct subscription_metadata {
-    struct aws_byte_buf topic;
-
-    napi_env env;
-
-    napi_ref on_suback;
-    napi_async_context on_suback_ctx;
-
-    napi_ref on_message;
-    napi_async_context on_message_ctx;
-};
-
-void s_on_subscribe_complete(
-    struct aws_mqtt_client_connection *connection,
-    uint16_t packet_id,
-    const struct aws_byte_cursor *topic,
-    enum aws_mqtt_qos qos,
-    int error_code,
-    void *userdata) {
-
-    (void)connection;
-
-    struct subscription_metadata *metadata = userdata;
-    napi_env env = metadata->env;
-
+static void s_dispatch_on_suback(struct mqtt_nodejs_callback_context *context) {
     napi_handle_scope handle_scope = NULL;
     napi_callback_scope cb_scope = NULL;
 
-    /* Don't need this anymore */
-    aws_byte_buf_clean_up(&metadata->topic);
-
-    /* Call callback */
-    if (metadata->on_suback) {
-
+    struct mqtt_nodejs_connection *node_connection = context->connection;
+    napi_env env = node_connection->env;
+    if (context->callback) {
         if (napi_open_handle_scope(env, &handle_scope)) {
+            s_raise_napi_error(env, s_handle_scope_open_failed);
             goto cleanup;
         }
 
         napi_value on_suback = NULL;
-        napi_get_reference_value(env, metadata->on_suback, &on_suback);
+        napi_get_reference_value(env, context->callback, &on_suback);
         if (on_suback) {
-
             napi_value resource_object = NULL;
             if (napi_create_object(env, &resource_object)) {
+                s_raise_napi_error(env, s_resource_creation_failed);
                 goto cleanup;
             }
 
-            if (napi_open_callback_scope(env, resource_object, metadata->on_suback_ctx, &cb_scope)) {
+            if (napi_open_callback_scope(env, resource_object, context->callback_ctx, &cb_scope)) {
+                s_raise_napi_error(env, s_callback_scope_open_failed);
                 goto cleanup;
             }
 
             napi_value recv;
             napi_value params[4];
-            if (napi_get_global(env, &recv) || napi_create_uint32(env, packet_id, &params[0]) ||
-                napi_create_string_utf8(env, (const char *)topic->ptr, topic->len, &params[1]) ||
-                napi_create_uint32(env, qos, &params[2]) || napi_create_int32(env, error_code, &params[3])) {
+            if (napi_get_global(env, &recv) || napi_create_int32(env, context->packet_id, &params[0]) ||
```

Comment:
Just to be sure, the callback scope will clean up all these temporaries when we close it, is that correct?

Reply:
This isn't particularly well documented in javascript, but the best I can tell is,

napi_create* creates a javascript object (identical to if you created it in javascript land) and will those be managed by javascript land when the user takes a reference to it.

https://nodejs.org/api/n-api.html#n_api_napi_create_object

Do you have any better thoughts?

Reply:
If you have more questions, I highly suggest reading this: https://v8.dev/docs/embed

the tl;dr on this operation tho is that creating a callback scope creates a HandleScope (on the heap), meaning all new objects exist in that scope. When the scope is destroyed, all handles are cleaned up (note: this does NOT mean the memory is released, that's up to the GC).